feat: new order

prathmeshkhandelwal1 opened the pull request:
# 🤖 Linear
Closes CONG-XXX

prathmeshkhandelwal1 committed:
feat: new order function

prathmeshkhandelwal1 commented:
DO NOT MERGE YET

prathmeshkhandelwal1 committed:
fix: refactor new_intent

prathmeshkhandelwal1 committed:
fix: build

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/intent/new_intent.rs:
```diff
@@ -161,10 +199,10 @@ pub fn new_intent(
 
     // Emit an event with full intent details.
-    emit_cpi!(IntentAddedEvent {
+    emit!(IntentAddedEvent {
```
need to use `emit_cpi!` here, in general we use `emit` for all event emission because `emit!` have a length limitation

harrier-lcc replied:
ah i see why you didnt use `emit_cpi` because of missing context. You need to either move this outside where context exist to call it, or expand the `emit_cpi` macro and replace the accounts used here

prathmeshkhandelwal1 replied:
fixed

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/intent/new_intent.rs:
```diff
@@ -38,9 +38,47 @@ pub fn new_intent(
     message_gas_limit: u64,
 ) -> Result<()> {
     // Clone to allow mut ref before move
```
remove this comment as well?

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/lib.rs:
```diff
@@ -87,6 +90,16 @@ pub mod everclear_spoke {
+        fee: u64,
+        deadline: i64,
+        sig: Vec<u8>,
+        params: Vec<OrderParameters>,
```
need to test the size limit of this, there is an 1232 size limit for each solana tx that this could hit

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/intent/new_intent.rs:
```diff
@@ -38,9 +38,47 @@ pub fn new_intent(
-    let state = &mut ctx.accounts.spoke_state;
+pub fn handle_new_intent<'info>(
+    accounts: &mut NewIntentAccounts<'info>,
+    program_id:Pubkey, // for ctx.programId
```
run `cargo fmt` and `cargo clippy` to lint it

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/intent/mod.rs:
```diff
@@ -127,7 +127,7 @@ fn u256_to_32bytes(val: impl Into<u128>) -> [u8; 32] {
 }
 
-fn u128_to_u256_be(val: u128) -> [u8; 32] {
+pub fn u128_to_u256_be(val: u128) -> [u8; 32] {
```
maybe move this to some utils.rs

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,90 @@
+    let mut intent_ids: Vec<[u8; 32]> = Vec::with_capacity(params.len());
+    for p in &params {
+        new_intent(
+            ctx.clone(),
```
how do this work? it havent change to `handle_new_intent` yet?

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,90 @@
+            p.message_gas_limit,
+        )?;
+
+        // Recompute intent ID exactly as in new_intent
```
we can make the `handle_new_intent` return the `intent_id`, this way you dont have to compute it again

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    // 4) Derive order ID and emit event
+    let order_id = compute_intent_hash(
```
how does this work? i think you need a separate hashing function here as its not intent that you are hashing (but intent id)?

harrier-lcc committed:
feat: ed25519 fee adapter signature verification

prathmeshkhandelwal1 committed:
fix: new_intent ctx-cpi

prathmeshkhandelwal1 committed:
fix: new_order method

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/types.rs:
```diff
@@ -0,0 +1,14 @@
+use anchor_lang::prelude::*;
+
+#[derive(AnchorSerialize, AnchorDeserialize, Clone)]
+pub struct OrderParameters {
```
better put this in, say `fee_adapter/new_order.rs` instead

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,106 @@
+    sig: Vec<u8>,
+    params: Vec<OrderParameters>,
+) -> Result<()> {
+    // 1) Validate inputs
```
have to add condition checks here (e.g. check if paused etc)
see the beginning of other entrypoints, e.g. `new_intent`

prathmeshkhandelwal1 replied:
all those checks have been moved to `handle_new_intent`, so it should take care of it

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,106 @@
+            destinations: p.destinations.clone(),
+            data: p.data.clone(),
+        };
+        let intent_id = compute_intent_hash(&evm_intent);
```
use the intent_id emitted by the event data instead of recomputing it

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,106 @@
+            p.data.clone(),
+            p.message_gas_limit,
+        );
+
```
also i think you have to `emit_cpi!(IntentAddedEvent)` here as well

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,106 @@
+        user: ctx.accounts.authority.key(),
+        intent_ids: intent_ids.clone(),
+        fee,
+        native_value: ctx.accounts.authority.lamports(),
```
`ctx.accounts.authority.lamports()` do not mean the SOL associated with the tx;

I think you need to pass an extra parameter for the function to indicate the fee in SOL instead

prathmeshkhandelwal1 committed:
fix: lamports converted to sol

prathmeshkhandelwal1 committed:
fix: handle fees native sol

harrier-lcc committed:
feat: solana new_intent with fee adapter logic

This integrate fee adapter logic in new intent call.

Note due to stack size limitation, the state in NewIntent are now
wrapped in Box to use heap space instead.

prathmeshkhandelwal1 committed:
fix: transaction size limit

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,119 @@
+
+    require!(!state.paused, SpokeError::ContractPaused);
+    require!(!params.is_empty(), SpokeError::EmptyParams);
+    require!(params.len() <= 6, SpokeError::TooManyOrders);
```
you can remove the hardcoded limit as if it hits the tx limit, it should error in rpc level instead of inside the contract

harrier-lcc commented on packages/contracts/solana-spoke/programs/everclear_spoke/src/instructions/fee_adapter/new_order.rs:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    let post_balance = **ctx.accounts.program_vault_account.to_account_info().lamports.borrow();
+    let native_value = post_balance.saturating_sub(pre_balance);
```
this part dont make sense now; the intent calls will not change the SOL balance for the program_vault_account

i think you need extra parameter to record this native_fee, see the other PR

prathmeshkhandelwal1 committed:
fix: merge new intent branch

prathmeshkhandelwal1 commented:
@harrier-lcc , I added the suggested changes, and also merged `handle_fees` in `new_order`, Also fixed this branch conflct with you branch, so it will be all good if we base it to your working branch `feat/solana-fees-adapter-new-intent`
